File: RBQFetchedResultsControllerExample/RBQFetchedResultsControllerExample/ExampleTableViewController.m
//
//  ExampleTableViewController.m
//  RBQFetchedResultsControllerExample
//
//  Created by Adam Fish on 1/5/15.
//  Copyright (c) 2015 Roobiq. All rights reserved.
//

#import "ExampleTableViewController.h"

#import "RBQFetchedResultsController.h"
#import "TestObject.h"
#import "RBQRealmNotificationManager.h"
#import "RLMRealm+Notifications.h"
#import "RLMObject+Notifications.h"

id NULL_IF_NIL(id x) {return x ? x : NSNull.null;}

@interface ExampleTableViewController () <RBQFetchedResultsControllerDelegate>

@property (strong, nonatomic) RBQFetchedResultsController *fetchedResultsController;

@end

@implementation ExampleTableViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    [realm beginWriteTransaction];
    
    [realm deleteAllObjects];
    
    for (NSUInteger i = 0; i < 1000; i++) {
        
        NSString *title = [NSString stringWithFormat:@"Cell %lu", (unsigned long)i];
        
        TestObject *object = [TestObject testObjectWithTitle:title sortIndex:i inTable:YES];
        
        if (i < 10) {
            object.sectionName = @"First Section";
        }
        else {
            object.sectionName = @"Second Section";
        }
        
        [realm addObject:object];
    }
    
    [realm commitWriteTransaction];
    
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"inTable = YES"];
    
    RBQFetchRequest *fetchRequest = [RBQFetchRequest fetchRequestWithEntityName:@"TestObject"
                                                                        inRealm:realm
                                                                      predicate:predicate];
    
    RLMSortDescriptor *sortDescriptor = [RLMSortDescriptor sortDescriptorWithProperty:@"sortIndex"
                                                                            ascending:YES];
    
    RLMSortDescriptor *sortDescriptorSection = [RLMSortDescriptor sortDescriptorWithProperty:@"sectionName"
                                                                                   ascending:YES];
    
    fetchRequest.sortDescriptors = @[sortDescriptorSection,sortDescriptor];
    
    self.fetchedResultsController = [[RBQFetchedResultsController alloc] initWithFetchRequest:fetchRequest
                                                                           sectionNameKeyPath:@"sectionName"
                                                                                    cacheName:@"testCache"];
    
    self.fetchedResultsController.delegate = self;
    
    [self.fetchedResultsController performFetch];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return [self.fetchedResultsController numberOfSections];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.fetchedResultsController numberOfRowsForSectionIndex:section];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return [self.fetchedResultsController titleForHeaderInSection:section];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"customCell" forIndexPath:indexPath];
    
    // Configure the cell...
    TestObject *objectForCell = [self.fetchedResultsController objectAtIndexPath:indexPath];
    
    cell.textLabel.text = objectForCell.title;
    
    
    return cell;
}

// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the specified item to be editable.
    return YES;
}

// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView
commitEditingStyle:(UITableViewCellEditingStyle)editingStyle
forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [self deleteObjectAtIndexPath:indexPath];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
        [self insertObject];
    }
}

#pragma mark - <RBQFetchedResultsControllerDelegate>

- (void)controllerWillChangeContent:(RBQFetchedResultsController *)controller
{
    NSLog(@"Beginning updates");
    [self.tableView beginUpdates];
}

- (void)controller:(RBQFetchedResultsController *)controller
   didChangeObject:(RBQSafeRealmObject *)anObject
       atIndexPath:(NSIndexPath *)indexPath
     forChangeType:(NSFetchedResultsChangeType)type
      newIndexPath:(NSIndexPath *)newIndexPath
{
    UITableView *tableView = self.tableView;
    
    switch(type) {
            
        case NSFetchedResultsChangeInsert:
        {
            NSLog(@"Inserting at path %@", newIndexPath);
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
        }
        case NSFetchedResultsChangeDelete:
        {
            NSLog(@"Deleting at path %ld", (long)indexPath.row);
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
        }
        case NSFetchedResultsChangeUpdate:
            NSLog(@"Updating at path %@", indexPath);
            if ([[tableView indexPathsForVisibleRows] containsObject:indexPath]) {
                [tableView reloadRowsAtIndexPaths:@[indexPath]
                                 withRowAnimation:UITableViewRowAnimationFade];
            }
            break;
            
        case NSFetchedResultsChangeMove:
            NSLog(@"Moving from path %@ to %@", indexPath, newIndexPath);
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
    }
}

- (void)controller:(RBQFetchedResultsController *)controller
  didChangeSection:(RBQFetchedResultsSectionInfo *)sectionInfo
           atIndex:(NSUInteger)sectionIndex
     forChangeType:(NSFetchedResultsChangeType)type
{
    UITableView *tableView = self.tableView;
    
    if (type == NSFetchedResultsChangeInsert) {
        NSLog(@"Inserting section at %lu", (unsigned long)sectionIndex);
        NSIndexSet *insertedSection = [NSIndexSet indexSetWithIndex:sectionIndex];
        
        [tableView insertSections:insertedSection withRowAnimation:UITableViewRowAnimationFade];
    }
    else if (type == NSFetchedResultsChangeDelete) {
        NSLog(@"Deleting section at %lu", (unsigned long)sectionIndex);
        NSIndexSet *deletedSection = [NSIndexSet indexSetWithIndex:sectionIndex];
        
        [tableView deleteSections:deletedSection withRowAnimation:UITableViewRowAnimationFade];
    }
}

- (void)controllerDidChangeContent:(RBQFetchedResultsController *)controller
{
    NSLog(@"Ending updates");
    NSLog(@"Fetched %ld Items After Change", (unsigned long)self.fetchedResultsController.fetchedObjects.count);
    @try {
        [self.tableView endUpdates];
    }
    @catch (NSException *ex) {
        NSLog(@"RBQFecthResultsTVC caught exception updating table view: %@. Falling back to reload.", ex);
        
        [self.fetchedResultsController reset];
        
        [self.tableView reloadData];
    }
}

#pragma mark - UIBarButton Actions

- (IBAction)didClickDeleteButton:(UIBarButtonItem *)sender
{
    // Delete the object in the first row
//    NSIndexPath *firstObjectIndexPath = [NSIndexPath indexPathForRow:0 inSection:0];
//    [self deleteObjectAtIndexPath:firstObjectIndexPath];
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        NSLog(@"DID BEGIN DELETE");
        
        // Test deleting a section (comment out above to test)
        RLMResults *objectInFirstSection = [TestObject objectsWhere:@"%K == %@",@"sectionName",@"First Section"];
        
        RLMRealm *realm = [RLMRealm defaultRealm];
        
        [realm beginWriteTransaction];
        [realm deleteObjectsWithNotification:objectInFirstSection];
        [realm commitWriteTransaction];
        
        NSLog(@"DID END DELETE");
    });
}

- (IBAction)didClickInsertButton:(UIBarButtonItem *)sender
{
    NSLog(@"DID BEGIN INSERT");
    NSLog(@"Fetched %ld Items Before Insert", (unsigned long)self.fetchedResultsController.fetchedObjects.count);
    [self insertObject];
    NSLog(@"DID END INSERT");
}

#pragma mark - Private

- (void)deleteObjectAtIndexPath:(NSIndexPath *)indexPath
{
    TestObject *object = [self.fetchedResultsController objectAtIndexPath:indexPath];
    if (!object) {
        return;
    }
    
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    [realm beginWriteTransaction];
    
    [realm deleteObjectWithNotification:object];
    
    [realm commitWriteTransaction];
}

- (void)insertObject
{
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    NSIndexPath *indexPathFirstRow = [NSIndexPath indexPathForRow:0 inSection:0];
    
    TestObject *object = [self.fetchedResultsController objectAtIndexPath:indexPathFirstRow];
    
    if (object.sortIndex > 0) {
        [realm beginWriteTransaction];
        
        NSInteger sortIndex = object.sortIndex - 1;
        
        NSString *title = [NSString stringWithFormat:@"Cell %lu", (unsigned long)sortIndex];
        
        
        TestObject *newObject = [TestObject objectInRealm:realm
                                            forPrimaryKey:[NSString stringWithFormat:@"%@%ld",title, (long)sortIndex]];
        
        if (!newObject) {
            newObject = [[TestObject alloc] init];
            newObject.title = title;
            newObject.sortIndex = sortIndex;
            newObject.sectionName = @"First Section";
            newObject.key = [NSString stringWithFormat:@"%@%ld",title, (long)sortIndex];
            newObject.inTable = YES;
            
            [realm addObjectWithNotification:newObject];
        }
        else {
            [newObject changeWithNotification:^(TestObject *testObject) {
                testObject.inTable = YES;
            }];
        }
        
        [realm commitWriteTransaction];
    }
    // Test Moves
    else {
        [realm beginWriteTransaction];
        
        NSIndexPath *indexPathFifthRow = [NSIndexPath indexPathForRow:5 inSection:0];
        NSIndexPath *indexPathThirdRow = [NSIndexPath indexPathForRow:3 inSection:0];
        NSIndexPath *indexPathSixthRow = [NSIndexPath indexPathForRow:6 inSection:0];
        NSIndexPath *indexPathFirstRow = [NSIndexPath indexPathForRow:0 inSection:0];
        
        TestObject *firstObject = [self.fetchedResultsController objectAtIndexPath:indexPathFirstRow];
        TestObject *thirdObject = [self.fetchedResultsController objectAtIndexPath:indexPathThirdRow];
        TestObject *fifthObject = [self.fetchedResultsController objectAtIndexPath:indexPathFifthRow];
        TestObject *sixthObject = [self.fetchedResultsController objectAtIndexPath:indexPathSixthRow];
        RLMResults *ninthObject = [TestObject objectsInRealm:realm where:@"%K == %@",@"title",@"Cell 9"];
        
        [fifthObject changeWithNotification:^(TestObject *testObject) {
            
            testObject.sortIndex += 1;
        }];
        
        [sixthObject changeWithNotification:^(TestObject *testObject) {
            
            testObject.sortIndex -= 1;
        }];
        
        [firstObject changeWithNotification:^(TestObject *testObject) {
            
            testObject.inTable = NO;
        }];
        
        [thirdObject changeWithNotification:^(TestObject *testObject) {
            
            testObject.title = @"Testing Move And Update";
        }];
        
        if (ninthObject.firstObject) {
            [ninthObject.firstObject changeWithNotification:^(TestObject *testObject) {
                
                if ([testObject.sectionName isEqualToString:@"First Section"]) {
                    testObject.sectionName = @"Second Section";
                }
                else {
                    testObject.sectionName = @"First Section";
                }
            }];
        }
        
         //Test an inserted section that's not first
//        TestObject *extraObjectInSection = [TestObject testObjectWithTitle:@"Test Section" sortIndex:3 inTable:YES];
//        extraObjectInSection.sectionName = @"Middle Section";
//        [realm addObject:extraObjectInSection];
//
//        [[RBQRealmNotificationManager defaultManager] didAddObjects:@[extraObjectInSection]
//                                                  willDeleteObjects:nil
//                                                   didChangeObjects:@[NULL_IF_NIL(fifthObject),
//                                                                      NULL_IF_NIL(sixthObject),
//                                                                      NULL_IF_NIL(firstObject),
//                                                                      NULL_IF_NIL(thirdObject)]];
        
        [realm commitWriteTransaction];
    }
}

@end

File: RBQFetchedResultsControllerSwiftExample/RBQFetchedResultsControllerSwiftExample/ExampleTableViewController.swift
//
//  ExampleTableViewController.swift
//  RBQFetchedResultsControllerSwiftExample
//
//  Created by Jeff on 4/2/15.
//  Copyright (c) 2015 Jeff. All rights reserved.
//

import UIKit
import Realm
import RBQFetchedResultsController


class ExampleTableViewController: UITableViewController, UITableViewDataSource, UITableViewDelegate, RBQFetchedResultsControllerDelegate {
    
    // MARK: variables
    let realm = RLMRealm.defaultRealm()
    var fetchedResultsController = RBQFetchedResultsController()
    
    func setUpFRC () {
        // Select and sort
        let predicate = NSPredicate(format: "inTable = YES")  // Can also use %i, %@
        let sortDescriptor = RLMSortDescriptor(property: "sortIndex", ascending: true)
        let sortDescriptorSection = RLMSortDescriptor(property: "sectionName", ascending: true)
        
        // Set up fetch
        let fetchRequest = RBQFetchRequest(entityName: "TestObject", inRealm: realm, predicate: predicate) // RBQFetchRequest(entityName: "InOut", inRealm: RLMRealm.defaultRealm(), predicate: predicate)
        fetchRequest.sortDescriptors = [sortDescriptor, sortDescriptorSection]  // can also be [primarySortDescriptor, secondarySortDescriptor]
        let fetchedResultsController = RBQFetchedResultsController(fetchRequest: fetchRequest, sectionNameKeyPath: "sectionName", cacheName: "testCache")
        
        fetchedResultsController.delegate = self
        let success = fetchedResultsController.performFetch()
        
        println("success: \(success), Total fetched: \(fetchedResultsController.fetchedObjects)")  //fetchedResultsController.fetchedObjects!.count
    }
    
    func initializeDB() {
        
        println ("realm.path: \(realm.path)")
    }
    
    
    
    
    // MARK: view
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        println ("realm.path: \(realm.path)")
        
        realm.beginWriteTransaction()
        realm.deleteAllObjects()
        
        for i in 0...1000 {
            let title = String("Cell \(i)")
//            let object = TestObject(title: title, sortIndex: i, inTable: true)
            let object = TestObject()
            object.title = title
            object.sortIndex = i
            object.inTable = true

            object.key = String(format: "\(object.title)\(object.sortIndex)")
            
            if i < 10 {
                object.sectionName = "First Section"
            } else {
                object.sectionName = "Second Section"
            }
            
            realm.addObject(object)
        }
        
        realm.commitWriteTransaction()
        
        // Set up fetch
        let predicate = NSPredicate(format: "inTable = YES")  // Can also use %i, %@
        let fetchRequest = RBQFetchRequest(entityName: "TestObject", inRealm: realm, predicate: predicate)
        
        let sortDescriptor = RLMSortDescriptor(property: "sortIndex", ascending: true)
        let sortDescriptorSection = RLMSortDescriptor(property: "sectionName", ascending: true)
        
        fetchRequest.sortDescriptors = [sortDescriptor, sortDescriptorSection]
        
        
        let fetchedResultsController = RBQFetchedResultsController(fetchRequest: fetchRequest, sectionNameKeyPath: "sectionName", cacheName: "testCache")
        
        fetchedResultsController.delegate = self
        let success = fetchedResultsController.performFetch()

    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    

    // MARK: Table view data source
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        let numberOfSections = fetchedResultsController.numberOfSections()
        return numberOfSections
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let numberOfRowsForSectionIndex = fetchedResultsController.numberOfRowsForSectionIndex(section)
        return fetchedResultsController.numberOfRowsForSectionIndex(section)
    }
    
    override func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let title = fetchedResultsController.titleForHeaderInSection(section)
        return title
    }
    
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        
        var cell = tableView.dequeueReusableCellWithIdentifier("customCell", forIndexPath: indexPath) as UITableViewCell
        
        let objectForCell = fetchedResultsController.objectInRealm(realm, atIndexPath: indexPath) as TestObject
        cell.textLabel?.text = objectForCell.title
        return cell
    }
    
    override func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        return true
    }
    
    override func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        if (editingStyle == UITableViewCellEditingStyle.Delete) {
            
            // Delete the row from the data source
            self.deleteObjectAtIndexPath(indexPath)
        } else if (editingStyle == UITableViewCellEditingStyle.Insert) {
            // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
            self.insertObject()
        }
    }
    

    // MARK: Private
    func deleteObjectAtIndexPath(indexPath : NSIndexPath) {
        if let object = fetchedResultsController.objectInRealm(realm, atIndexPath: indexPath) as? TestObject {
            realm.beginWriteTransaction()
            realm.deleteObjectWithNotification(object)
            realm.commitWriteTransaction()
        }
    }
    
    func insertObject() {
        let indexPathFirstRow =  NSIndexPath(forRow: 0, inSection: 0)
        
        let object = fetchedResultsController.objectInRealm(realm, atIndexPath: indexPathFirstRow) as TestObject
        
        if object.sortIndex > 0 {
            realm.beginWriteTransaction()
            
            let sortIndex = object.sortIndex - 1;
            
            let title = NSString(format: "Cell %lu", CLong(sortIndex))
            
            if let newObject = RLMObject(inRealm: realm, forPrimaryKey: NSString(format: "%@%ld", title, CLong(sortIndex))) {
                newObject.changeWithNotification{ object in
                    let testObject = object as TestObject
                    testObject.inTable = true
                }
            } else {
//                let newObject = TestObject(title: title, sortIndex: sortIndex, inTable: true)
                let newObject = TestObject()
                newObject.title = title
                newObject.sortIndex = sortIndex
                newObject.inTable = true
                newObject.sectionName = "First Section"
                newObject.key = NSString(format: "%@%ld", title, CLong(sortIndex))
                realm.addObjectWithNotification(newObject)
            }
            realm.commitWriteTransaction()
        }
        

//    // Test Moves
//    else {
//    [realm beginWriteTransaction];
//    
//    NSIndexPath *indexPathFifthRow = [NSIndexPath indexPathForRow:5 inSection:0];
//    NSIndexPath *indexPathThirdRow = [NSIndexPath indexPathForRow:3 inSection:0];
//    NSIndexPath *indexPathSixthRow = [NSIndexPath indexPathForRow:6 inSection:0];
//    NSIndexPath *indexPathFirstRow = [NSIndexPath indexPathForRow:0 inSection:0];
//    
//    TestObject *firstObject = [self.fetchedResultsController objectInRealm:realm
//    atIndexPath:indexPathFirstRow];
//    TestObject *thirdObject = [self.fetchedResultsController objectInRealm:realm
//    atIndexPath:indexPathThirdRow];
//    TestObject *fifthObject = [self.fetchedResultsController objectInRealm:realm
//    atIndexPath:indexPathFifthRow];
//    TestObject *sixthObject = [self.fetchedResultsController objectInRealm:realm
//    atIndexPath:indexPathSixthRow];
//    RLMResults *ninthObject = [TestObject objectsInRealm:realm where:@"%K == %@",@"title",@"Cell 9"];
//    
//    [fifthObject changeWithNotification:^(RLMObject *object) {
//    TestObject *testObject = (TestObject *)object;
//    
//    testObject.sortIndex += 1;
//    }];
//    
//    [sixthObject changeWithNotification:^(RLMObject *object) {
//    TestObject *testObject = (TestObject *)object;
//    
//    testObject.sortIndex -= 1;
//    }];
//    
//    [firstObject changeWithNotification:^(RLMObject *object) {
//    TestObject *testObject = (TestObject *)object;
//    
//    testObject.inTable = NO;
//    }];
//    
//    [thirdObject changeWithNotification:^(RLMObject *object) {
//    TestObject *testObject = (TestObject *)object;
//    
//    testObject.title = @"Testing Move And Update";
//    }];
//    
//    if (ninthObject.firstObject) {
//    [ninthObject.firstObject changeWithNotification:^(RLMObject *object) {
//    TestObject *testObject = (TestObject *)object;
//    
//    if ([testObject.sectionName isEqualToString:@"First Section"]) {
//    testObject.sectionName = @"Second Section";
//    }
//    else {
//    testObject.sectionName = @"First Section";
//    }
//    }];
//    }
//    
//    //Test an inserted section that's not first
//    //        TestObject *extraObjectInSection = [TestObject testObjectWithTitle:@"Test Section" sortIndex:3 inTable:YES];
//    //        extraObjectInSection.sectionName = @"Middle Section";
//    //        [realm addObject:extraObjectInSection];
//    //
//    //        [[RBQRealmNotificationManager defaultManager] didAddObjects:@[extraObjectInSection]
//    //                                                  willDeleteObjects:nil
//    //                                                   didChangeObjects:@[NULL_IF_NIL(fifthObject),
//    //                                                                      NULL_IF_NIL(sixthObject),
//    //                                                                      NULL_IF_NIL(firstObject),
//    //                                                                      NULL_IF_NIL(thirdObject)]];
//    
//    [realm commitWriteTransaction];
//    }
//    }
    
   }
    
}